setup.py

```diff
@@ -2447,24 +2447,40 @@ def split_var(name, sep):
         else:
             runtime_library_dirs = [openssl_rpath]
 
+        kw = dict(
+            include_dirs=openssl_includes,
+            library_dirs=openssl_libdirs,
+            libraries=openssl_libs,
+            runtime_library_dirs=runtime_library_dirs,
+        )
+        # Don't re-export OpenSSL symbols from statically linked libs if
+        # OpenSSL is built with "./config no-shared -fPIC".
+        # Note: gcc -shared -lcrypto -L/path/to/openssl/lib automatically
+        # falls back to libcrypto.a if the library search directories do not
+        # contain libcrypto.so.
+        if sysconfig.get_config_var("PY_LINKER_EXCLUDE_LIBS"):
+            kw["extra_link_args"] = [
+                f"-Wl,--exclude-libs,lib{lib}.a" for lib in kw["libraries"]
+            ]
+
         if config_vars.get("HAVE_X509_VERIFY_PARAM_SET1_HOST"):
-            self.add(Extension(
-                '_ssl', ['_ssl.c'],
-                include_dirs=openssl_includes,
-                library_dirs=openssl_libdirs,
-                libraries=openssl_libs,
-                runtime_library_dirs=runtime_library_dirs,
-                depends=['socketmodule.h', '_ssl/debughelpers.c'])
+            self.add(
+                Extension(
+                    '_ssl', ['_ssl.c'],
```

Comment:
Instead on reliying on the absence of the shared library for openssl, we should force the linker to pick the archive. The syntax is:

```
-l:ssl.a
```

this needs to be added to `extra_link_args` parameter and drop the one in `libraries`

Reply:
I'm strong -1. This feature is already beyond what I originally wanted to commit to.

Your request would make things even more complicated. I really don't want to support static linking because I have neither time nor motivation to deal with the additional support overhead. Static linking of OpenSSL has negative consequences for security: you cannot easily update.

PyCA/cryptography deals with update issue of static-linked code by tying the release cycle to OpenSSL security releases. Alex and Paul release a new version a couple of hours after each and every OpenSSL security release. Rhetorical question: As the new RM do you want to commit to the same release policy? (please say: "no").

Reply:
Agreed, the goal of this PR isn't to static link by default.  It's just to enable the possibility of doing that in a reasonable isolated manner when someone builds against a static only OpenSSL.

Reply:
> As the new RM do you want to commit to the same release policy? (please say: "no").

I say "no" :)

> I'm strong -1. This feature is already beyond what I originally wanted to commit to.

If you are -1, let's not do it. I trust your juzguement here.

> Your request would make things even more complicated. I really don't want to support static linking because I have neither time nor motivation to deal with the additional support overhead. Static linking of OpenSSL has negative consequences for security: you cannot easily update.

I think also that there is some misunderstanding in what I am trying to suggest here. I will try to explain it better:

* I don't want us to officially support static linking if we don't want.
* I don't want to statically link by default
* I don't want us to support many ways to do things

What I want here is to offer some safe enough alternative to those users that need to download their own version of OpenSSL in an environment where OpenSSL 1.0.2 is present. We don't need to officially support it, but I think we need to offer a way so this can be achieved without having to patch CPython. The reason this may be more important is due to the fact that shortly after 3.10 is released 3.9 will lose bugfix support so users in this situation will have no way to use a bugfix-supported Python.

I don't have special interest in pushing one solution or the other: I am triying to explain the concerns that exist in the general ecosystem. You are our trusted expert in SSL and security and we trust you and your juzguement.

Please, understand that my comments are suggestions and general questions so we can provide a good experience for everyone, they are not requirements or something you need to act on now :)

Reply:
> Agreed, the goal of this PR isn't to static link by default. It's just to enable the possibility of doing that in a reasonable isolated manner when someone builds against a static only OpenSSL.

Right, I think I have not expressed myself correctly. What I was suggesting is to offer an *alternative, opt-in* way to statically compile openssl

Reply:
> What I want here is to offer some safe enough alternative to those users that need to download their own version of OpenSSL in an environment where OpenSSL 1.0.2 is present. We don't need to officially support it, but I think we need to offer a way so this can be achieved without having to patch CPython.

I believe this PR accomplishes that?  We give people instructions on how to do the no-shared build of OpenSSL and how to point CPython's configure at that and they should be good to go without any consequences of interactions with the system's dynamic libssl that other extension modules they have might load.

Reply:
> I believe this PR accomplishes that?

Yeah, but as I understand it only works if you don't build the shared object for OpenSSL and therefore `libssl.so` is not on the link search path. On the other hand, using `-l:ssl.a` (only when we opt-in, not by default) will work in all cases.

Reply:
How would you implement it? An opt-in would require a autoconf flag, which would then codify official support for the feature.

Reply:
We could:

* Configure option that explicitly states that is not an officially supported option in the help section.

* Environment variable present that we check in the `setup.py`>

* Maybe some specific flag in the `pyconfig.h`

In general terms we either don't document it directly or (my preference) we explicitly state that whatever we use is not officially supported and it's only purpose is to make life easier for specific use cases and that users are responsible of making sure everything works correctly.

Reply:
How about ``PY_UNSUPPORTED_OPENSSL_BUILD=static make``?

By the way your instructions are incorrect. :)

Reply:
> How about ``PY_UNSUPPORTED_OPENSSL_BUILD=static make`

👍

> By the way your instructions are incorrect. :)

Oh! What didn't work? I made a rough prototype with my instructions and that worked on my side. What errors are you experiencing?

Reply:
``-l:ssl.a`` does not work for me (GCC 10). It has to be filename of the archive: ``-l:libssl.a``

Reply:
> `-l:ssl.a` does not work for me (GCC 10). It has to be filename of the archive: `-l:libssl.a`

Oh, right, that was totally my fault. It should have said `-l:libssl.a` indeed :)